File: src/rlm_research_planner/services/speedup_inventory.py
from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from math import ceil, gcd
from typing import Iterable

from rlm_research_planner.domain.models import (
    PaidItem,
    PaidOffer,
    SpeedupInventoryItem,
)
from rlm_research_planner.services.paid_value import (
    minimum_gems_for_speedup_seconds,
)


SPEEDUP_KINDS = (
    "general",
    "research",
    "training",
    "construction",
    "healing",
    "merging",
    "crafting",
)


@dataclass(frozen=True)
class SpeedupCoverage:
    target_kind: str
    required_seconds: int
    available_seconds: int
    applied_seconds: int
    remaining_seconds: int
    surplus_seconds: int
    remaining_task_seconds: tuple[int, ...] = ()
    used_items: tuple[SpeedupInventoryItem, ...] = ()


@dataclass(frozen=True)
class PaidOfferRecommendation:
    offer_id: str
    title: str
    purchases: int
    seconds_per_purchase: int
    total_seconds: int
    diamond_cost_each: int
    total_diamond_cost: int | None
    gems_per_purchase: int
    available_gems: int
    applied_speedup_seconds: int
    gems_used: int
    gem_applied_seconds: int
    remaining_seconds: int
    excess_seconds: int


def normalize_speedup_inventory(
    entries: Iterable[SpeedupInventoryItem],
) -> tuple[SpeedupInventoryItem, ...]:
    totals: dict[tuple[str, int], int] = {}
    for entry in entries:
        kind = str(entry.kind).strip()
        duration = max(0, int(entry.duration_seconds))
        quantity = max(0, int(entry.quantity))
        if kind not in SPEEDUP_KINDS or duration <= 0 or quantity <= 0:
            continue
        key = (kind, duration)
        totals[key] = totals.get(key, 0) + quantity
    order = {kind: index for index, kind in enumerate(SPEEDUP_KINDS)}
    return tuple(
        SpeedupInventoryItem(kind, duration, quantity)
        for (kind, duration), quantity in sorted(
            totals.items(),
            key=lambda item: (order[item[0][0]], -item[0][1]),
        )
    )


def applicable_speedup_seconds(
    entries: Iterable[SpeedupInventoryItem],
    target_kind: str,
) -> int:
    eligible = {"general", target_kind}
    return sum(
        max(0, int(entry.duration_seconds)) * max(0, int(entry.quantity))
        for entry in entries
        if entry.kind in eligible
    )


def _normalized_task_seconds(
    required_seconds: int,
    task_seconds: Iterable[int] | None,
) -> list[int]:
    required = max(0, int(required_seconds))
    if task_seconds is None:
        return [required] if required else []
    tasks = [max(0, int(value)) for value in task_seconds if int(value) > 0]
    known_total = sum(tasks)
    if known_total < required:
        tasks.append(required - known_total)
    return tasks


def _allocate_without_overrun(
    entries: Iterable[SpeedupInventoryItem],
    target_kind: str,
    task_seconds: Iterable[int],
) -> tuple[tuple[int, ...], tuple[SpeedupInventoryItem, ...]]:
    remaining = [max(0, int(value)) for value in task_seconds if int(value) > 0]
    eligible = {"general", target_kind}
    candidates = list(
        sorted(
            (
                entry
                for entry in normalize_speedup_inventory(entries)
                if entry.kind in eligible
            ),
            key=lambda entry: (
                -entry.duration_seconds,
                entry.kind == "general",
            ),
        )
    )
    # A duration can never be used more often than the sum of the per-task
    # capacities.  Paid-offer simulations may multiply an offer thousands of
    # times while looking for a completing purchase count; retaining those
    # impossible excess quantities makes the subset solver needlessly huge.
    quantities = [
        min(
            max(0, int(entry.quantity)),
            sum(task // entry.duration_seconds for task in remaining),
        )
        for entry in candidates
    ]
    used_quantities = [0] * len(candidates)

    for task_index in sorted(range(len(remaining)), key=remaining.__getitem__):
        task = remaining[task_index]
        available_indexes = [
            index
            for index, entry in enumerate(candidates)
            if quantities[index] > 0 and 0 < entry.duration_seconds <= task
        ]
        if not available_indexes:
            continue
        scale = 0
        for index in available_indexes:
            scale = gcd(scale, candidates[index].duration_seconds)
        capacity = task // max(1, scale)
        chunks: list[tuple[int, int, int]] = []
        for index in available_indexes:
            remaining_quantity = quantities[index]
            chunk_quantity = 1
            unit = candidates[index].duration_seconds // scale
            while remaining_quantity > 0:
                count = min(chunk_quantity, remaining_quantity)
                chunks.append((index, count, unit * count))
                remaining_quantity -= count
                chunk_quantity *= 2

        selected_counts: dict[int, int] = {}
        if capacity <= 2_000_000:
            mask = (1 << (capacity + 1)) - 1
            reachable = 1
            newly_reachable: list[int] = []
            for _index, _count, weight in chunks:
                new_states = ((reachable << weight) & mask) & ~reachable
                newly_reachable.append(new_states)
                reachable |= new_states
            target = reachable.bit_length() - 1
            selected = target
            for chunk_index in range(len(chunks) - 1, -1, -1):
                if (newly_reachable[chunk_index] >> selected) & 1:
                    entry_index, count, weight = chunks[chunk_index]
                    selected_counts[entry_index] = (
                        selected_counts.get(entry_index, 0) + count
                    )
                    selected -= weight
            applied = target * scale
        else:
            applied = 0
            for index in available_indexes:
                duration = candidates[index].duration_seconds
                count = min(quantities[index], (task - applied) // duration)
                if count > 0:
                    selected_counts[index] = count
                    applied += duration * count

        remaining[task_index] -= applied
        for index, count in selected_counts.items():
            quantities[index] -= count
            used_quantities[index] += count

    used = tuple(
        SpeedupInventoryItem(entry.kind, entry.duration_seconds, count)
        for entry, count in zip(candidates, used_quantities, strict=True)
        if count > 0
    )
    return tuple(remaining), used


@lru_cache(maxsize=4096)
def _unlimited_exact_duration(
    seconds: int,
    durations: tuple[int, ...],
) -> bool:
    """Whether unlimited items can fill one task without exceeding it.

    This inexpensive feasibility gate prevents paid-offer recommendation from
    repeatedly running the bounded allocator when no number of purchases can
    ever fill a task exactly (for example, only five-minute items for a timer
    with a non-five-minute remainder).
    """

    target = max(0, int(seconds))
    normalized = tuple(sorted({max(0, int(value)) for value in durations} - {0}))
    if target <= 0:
        return True
    if not normalized:
        return False
    scale = 0
    for duration in normalized:
        scale = gcd(scale, duration)
    scale = max(1, scale)
    if target % scale:
        return False
    target //= scale
    units = tuple(duration // scale for duration in normalized)
    if 1 in units:
        return True
    reachable = bytearray(target + 1)
    reachable[0] = 1
    for value in range(target + 1):
        if not reachable[value]:
            continue
        for unit in units:
            next_value = value + unit
            if next_value <= target:
                reachable[next_value] = 1
    return bool(reachable[target])


def speedup_coverage(
    required_seconds: int,
    entries: Iterable[SpeedupInventoryItem],
    target_kind: str,
    task_seconds: Iterable[int] | None = None,
) -> SpeedupCoverage:
    tasks = _normalized_task_seconds(required_seconds, task_seconds)
    required = sum(tasks)
    normalized_entries = normalize_speedup_inventory(entries)
    available = applicable_speedup_seconds(normalized_entries, target_kind)
    remaining_tasks, used_items = _allocate_without_overrun(
        normalized_entries,
        target_kind,
        tasks,
    )
    remaining = sum(remaining_tasks)
    applied = max(0, required - remaining)
    return SpeedupCoverage(
        target_kind=target_kind,
        required_seconds=required,
        available_seconds=available,
        applied_seconds=applied,
        remaining_seconds=remaining,
        surplus_seconds=max(0, available - applied),
        remaining_task_seconds=remaining_tasks,
        used_items=used_items,
    )


def add_paid_items_to_inventory(
    entries: Iterable[SpeedupInventoryItem],
    paid_items: Iterable[PaidItem],
) -> tuple[SpeedupInventoryItem, ...]:
    additions = [
        SpeedupInventoryItem(
            kind=item.kind,
            duration_seconds=max(0, int(item.duration_seconds)),
            quantity=max(0, int(item.quantity)),
        )
        for item in paid_items
        if item.kind in SPEEDUP_KINDS
    ]
    return normalize_speedup_inventory((*entries, *additions))


def paid_offer_speedup_seconds(offer: PaidOffer, target_kind: str) -> int:
    eligible = {"general", target_kind}
    return sum(
        max(0, int(item.duration_seconds)) * max(0, int(item.quantity))
        for item in offer.items
        if item.kind in eligible
    )


def paid_offer_gems(offer: PaidOffer) -> int:
    return (
        max(0, int(offer.included_gems))
        + max(0, int(offer.bonus_gems))
        + sum(
            max(0, int(item.quantity))
            for item in offer.items
            if item.kind == "gems"
        )
    )


def _offer_covers_shortfall(
    shortfall_seconds: int,
    offer: PaidOffer,
    target_kind: str,
    gems_per_purchase: int,
    purchases: int,
    task_seconds: tuple[int, ...],
    use_gems: bool,
) -> bool:
    repeated_items = tuple(
        PaidItem(
            kind=item.kind,
            name=item.name,
            quantity=max(0, int(item.quantity)) * purchases,
            duration_seconds=item.duration_seconds,
            gem_value_each=item.gem_value_each,
            points_each=item.points_each,
        )
        for item in offer.items
    )
    coverage = speedup_coverage(
        shortfall_seconds,
        (
            SpeedupInventoryItem(
                item.kind,
                max(0, int(item.duration_seconds)),
                max(0, int(item.quantity)),
            )
            for item in repeated_items
            if item.kind in SPEEDUP_KINDS
        ),
        target_kind,
        task_seconds,
    )
    if coverage.remaining_seconds <= 0:
        return True
    if not use_gems or gems_per_purchase <= 0:
        return False
    required_gems = sum(
        minimum_gems_for_speedup_seconds(seconds).gems
        for seconds in coverage.remaining_task_seconds
    )
    return gems_per_purchase * purchases >= required_gems


def _minimum_offer_purchases(
    shortfall_seconds: int,
    offer: PaidOffer,
    target_kind: str,
    gems_per_purchase: int,
    task_seconds: tuple[int, ...],
    use_gems: bool,
) -> int | None:
    eligible_durations = [
        max(0, int(item.duration_seconds))
        for item in offer.items
        if item.kind in {"general", target_kind}
        and int(item.duration_seconds) > 0
        and int(item.quantity) > 0
    ]
    if (
        (not use_gems or gems_per_purchase <= 0)
        and eligible_durations
        and any(
            not _unlimited_exact_duration(seconds, tuple(eligible_durations))
            for seconds in task_seconds
        )
    ):
        return None
    if use_gems and gems_per_purchase > 0:
        gems_without_speedups = sum(
            minimum_gems_for_speedup_seconds(seconds).gems
            for seconds in task_seconds
        )
        high = max(1, ceil(gems_without_speedups / gems_per_purchase))
    elif eligible_durations:
        minimum_duration = min(eligible_durations)
        high = max(
            1,
            sum(ceil(seconds / minimum_duration) for seconds in task_seconds),
        )
    else:
        return None
    if not _offer_covers_shortfall(
        shortfall_seconds,
        offer,
        target_kind,
        gems_per_purchase,
        high,
        task_seconds,
        use_gems,
    ):
        return None
    low = 1
    while low < high:
        middle = (low + high) // 2
        if _offer_covers_shortfall(
            shortfall_seconds,
            offer,
            target_kind,
            gems_per_purchase,
            middle,
            task_seconds,
            use_gems,
        ):
            high = middle
        else:
            low = middle + 1
    return low


def recommend_paid_offers(
    shortfall_seconds: int,
    offers: Iterable[PaidOffer],
    target_kind: str,
    *,
    limit: int = 3,
    task_seconds: Iterable[int] | None = None,
    use_gems: bool = False,
) -> tuple[PaidOfferRecommendation, ...]:
    tasks = tuple(_normalized_task_seconds(shortfall_seconds, task_seconds))
    shortfall = sum(tasks)
    if shortfall <= 0:
        return ()
    recommendations: list[PaidOfferRecommendation] = []
    for offer in offers:
        seconds = paid_offer_speedup_seconds(offer, target_kind)
        gems = paid_offer_gems(offer)
        if seconds <= 0 and (not use_gems or gems <= 0):
            continue
        purchases = _minimum_offer_purchases(
            shortfall,
            offer,
            target_kind,
            gems,
            tasks,
            use_gems,
        )
        # A saved offer is still useful even when repeated purchases cannot
        # finish every task without wasting a speed-up item.  In that case,
        # show the contribution from one purchase and the remaining time
        # instead of hiding the offer entirely.
        if purchases is None:
            purchases = 1
        total_seconds = seconds * purchases
        available_gems = gems * purchases
        repeated_inventory = tuple(
            SpeedupInventoryItem(
                item.kind,
                max(0, int(item.duration_seconds)),
                max(0, int(item.quantity)) * purchases,
            )
            for item in offer.items
            if item.kind in SPEEDUP_KINDS
        )
        coverage = speedup_coverage(
            shortfall,
            repeated_inventory,
            target_kind,
            tasks,
        )
        gem_purchases = tuple(
            minimum_gems_for_speedup_seconds(value)
            for value in coverage.remaining_task_seconds
        )
        required_gems = sum(item.gems for item in gem_purchases)
        can_use_gems = (
            use_gems
            and coverage.remaining_seconds > 0
            and available_gems >= required_gems
        )
        if coverage.applied_seconds <= 0 and not can_use_gems:
            continue
        gems_used = required_gems if can_use_gems else 0
        gem_applied_seconds = coverage.remaining_seconds if can_use_gems else 0
        remaining_seconds = max(
            0,
            coverage.remaining_seconds - gem_applied_seconds,
        )
        cost_each = max(0, int(offer.diamond_cost))
        recommendations.append(
            PaidOfferRecommendation(
                offer_id=offer.offer_id,
                title=offer.title,
                purchases=purchases,
                seconds_per_purchase=seconds,
                total_seconds=total_seconds,
                diamond_cost_each=cost_each,
                total_diamond_cost=(cost_each * purchases if cost_each else None),
                gems_per_purchase=gems,
                available_gems=available_gems,
                applied_speedup_seconds=coverage.applied_seconds,
                gems_used=gems_used,
                gem_applied_seconds=gem_applied_seconds,
                remaining_seconds=remaining_seconds,
                excess_seconds=max(
                    0,
                    coverage.surplus_seconds
                    + sum(item.purchased_seconds for item in gem_purchases)
                    - gem_applied_seconds,
                ),
            )
        )
    recommendations.sort(
        key=lambda item: (
            item.remaining_seconds > 0,
            item.total_diamond_cost is None,
            item.total_diamond_cost if item.total_diamond_cost is not None else 0,
            item.excess_seconds,
            item.purchases,
            item.title.casefold(),
        )
    )
    return tuple(recommendations[: max(0, int(limit))])

File: tools/RLMResearchPlannerPWA/src/speedup-inventory.js
import { minimumGemsForSpeedupSeconds } from "./paid-value.js?v=0.1.7-b1";

export const SPEEDUP_KINDS = [
  "general", "research", "training", "construction", "healing", "merging", "crafting",
];

export const SPEEDUP_DURATION_SECONDS = [
  60, 3 * 60, 5 * 60, 10 * 60, 15 * 60, 30 * 60, 60 * 60,
  3 * 60 * 60, 8 * 60 * 60, 15 * 60 * 60, 24 * 60 * 60,
  3 * 24 * 60 * 60, 7 * 24 * 60 * 60, 30 * 24 * 60 * 60,
];

export const SPEEDUP_DURATION_GROUPS = [
  ["minutes", SPEEDUP_DURATION_SECONDS.slice(0, 7)],
  ["hours", SPEEDUP_DURATION_SECONDS.slice(7, 11)],
  ["days", SPEEDUP_DURATION_SECONDS.slice(11)],
];

export function speedupDurationGroup(durationSeconds) {
  const seconds = Math.max(0, Math.trunc(Number(durationSeconds) || 0));
  if (seconds <= 60 * 60) return "minutes";
  if (seconds <= 24 * 60 * 60) return "hours";
  return "days";
}

const integer = (value) => Math.max(0, Math.trunc(Number(value) || 0));

export function normalizeSpeedupInventory(entries) {
  const totals = new Map();
  for (const raw of Array.isArray(entries) ? entries : []) {
    const kind = String(raw?.kind || "general");
    const durationSeconds = integer(raw?.durationSeconds ?? raw?.duration_seconds);
    const quantity = integer(raw?.quantity);
    if (!SPEEDUP_KINDS.includes(kind) || !durationSeconds || !quantity) continue;
    const key = `${kind}\u0000${durationSeconds}`;
    totals.set(key, (totals.get(key) || 0) + quantity);
  }
  return [...totals.entries()]
    .map(([key, quantity]) => {
      const [kind, duration] = key.split("\u0000");
      return { kind, durationSeconds: Number(duration), quantity };
    })
    .sort((left, right) => SPEEDUP_KINDS.indexOf(left.kind) - SPEEDUP_KINDS.indexOf(right.kind)
      || right.durationSeconds - left.durationSeconds);
}

export function saveSpeedupInventoryEntry(entries, index, entry) {
  const normalized = normalizeSpeedupInventory(entries);
  const targetIndex = Number.isInteger(index) ? index : -1;
  if (targetIndex >= 0 && targetIndex < normalized.length) normalized.splice(targetIndex, 1, entry);
  else normalized.push(entry);
  return normalizeSpeedupInventory(normalized);
}

export function deleteSpeedupInventoryEntry(entries, index) {
  const normalized = normalizeSpeedupInventory(entries);
  const targetIndex = Number.isInteger(index) ? index : -1;
  if (targetIndex >= 0 && targetIndex < normalized.length) normalized.splice(targetIndex, 1);
  return normalized;
}

export function applicableSpeedupSeconds(entries, targetKind) {
  const eligible = new Set(["general", targetKind]);
  return normalizeSpeedupInventory(entries).reduce(
    (total, entry) => total + (eligible.has(entry.kind) ? entry.durationSeconds * entry.quantity : 0),
    0,
  );
}

function normalizedTaskSeconds(requiredSeconds, taskSeconds) {
  const required = integer(requiredSeconds);
  if (!Array.isArray(taskSeconds)) return required ? [required] : [];
  const tasks = taskSeconds.map(integer).filter((value) => value > 0);
  const knownTotal = tasks.reduce((sum, value) => sum + value, 0);
  if (knownTotal < required) tasks.push(required - knownTotal);
  return tasks;
}

function greatestCommonDivisor(left, right) {
  let a = Math.abs(integer(left));
  let b = Math.abs(integer(right));
  while (b) [a, b] = [b, a % b];
  return a;
}

function allocateWithoutOverrun(entries, targetKind, taskSeconds) {
  const remainingTaskSeconds = taskSeconds.map(integer).filter((value) => value > 0);
  const eligible = new Set(["general", targetKind]);
  const candidates = normalizeSpeedupInventory(entries)
    .filter((entry) => eligible.has(entry.kind))
    .sort((left, right) => right.durationSeconds - left.durationSeconds
      || Number(left.kind === "general") - Number(right.kind === "general"));
  // Ignore quantities that cannot fit into the individual tasks. Paid-offer
  // simulations can otherwise create enormous, unusable inventories while
  // searching for a purchase count.
  const quantities = candidates.map((entry) => Math.min(
    entry.quantity,
    remainingTaskSeconds.reduce(
      (total, seconds) => total + Math.floor(seconds / entry.durationSeconds),
      0,
    ),
  ));
  const usedQuantities = candidates.map(() => 0);
  const taskOrder = remainingTaskSeconds
    .map((seconds, index) => ({ seconds, index }))
    .sort((left, right) => left.seconds - right.seconds || left.index - right.index);

  for (const { index: taskIndex } of taskOrder) {
    const task = remainingTaskSeconds[taskIndex];
    const availableIndexes = candidates
      .map((entry, index) => ({ entry, index }))
      .filter(({ entry, index }) => quantities[index] > 0
        && entry.durationSeconds > 0
        && entry.durationSeconds <= task)
      .map(({ index }) => index);
    if (!availableIndexes.length) continue;
    const scale = availableIndexes.reduce(
      (value, index) => greatestCommonDivisor(value, candidates[index].durationSeconds),
      0,
    ) || 1;
    const capacity = Math.floor(task / scale);
    const chunks = [];
    for (const index of availableIndexes) {
      let remainingQuantity = quantities[index];
      let chunkQuantity = 1;
      const unit = Math.floor(candidates[index].durationSeconds / scale);
      while (remainingQuantity > 0) {
        const count = Math.min(chunkQuantity, remainingQuantity);
        chunks.push({ index, count, weight: unit * count });
        remainingQuantity -= count;
        chunkQuantity *= 2;
      }
    }

    const selectedCounts = new Map();
    let applied = 0;
    if (capacity <= 500_000) {
      const mask = (1n << BigInt(capacity + 1)) - 1n;
      let reachable = 1n;
      const newlyReachable = [];
      for (const chunk of chunks) {
        const states = ((reachable << BigInt(chunk.weight)) & mask) & ~reachable;
        newlyReachable.push(states);
        reachable |= states;
      }
      const target = reachable.toString(2).length - 1;
      let selected = target;
      for (let chunkIndex = chunks.length - 1; chunkIndex >= 0; chunkIndex -= 1) {
        if (((newlyReachable[chunkIndex] >> BigInt(selected)) & 1n) === 0n) continue;
        const chunk = chunks[chunkIndex];
        selectedCounts.set(chunk.index, (selectedCounts.get(chunk.index) || 0) + chunk.count);
        selected -= chunk.weight;
      }
      applied = target * scale;
    } else {
      for (const index of availableIndexes) {
        const duration = candidates[index].durationSeconds;
        const count = Math.min(quantities[index], Math.floor((task - applied) / duration));
        if (!count) continue;
        selectedCounts.set(index, count);
        applied += duration * count;
      }
    }
    remainingTaskSeconds[taskIndex] -= applied;
    for (const [index, count] of selectedCounts) {
      quantities[index] -= count;
      usedQuantities[index] += count;
    }
  }
  const usedItems = candidates
    .map((entry, index) => ({ ...entry, quantity: usedQuantities[index] }))
    .filter((entry) => entry.quantity > 0);
  return { remainingTaskSeconds, usedItems };
}

function hasUnlimitedExactDuration(seconds, durations) {
  let target = integer(seconds);
  const normalized = [...new Set(durations.map(integer).filter((value) => value > 0))]
    .sort((left, right) => left - right);
  if (!target) return true;
  if (!normalized.length) return false;
  const scale = normalized.reduce(greatestCommonDivisor, 0) || 1;
  if (target % scale) return false;
  target = Math.floor(target / scale);
  const units = normalized.map((duration) => Math.floor(duration / scale));
  if (units.includes(1)) return true;
  const reachable = new Uint8Array(target + 1);
  reachable[0] = 1;
  for (let value = 0; value <= target; value += 1) {
    if (!reachable[value]) continue;
    for (const unit of units) {
      const next = value + unit;
      if (next <= target) reachable[next] = 1;
    }
  }
  return reachable[target] === 1;
}

export function speedupCoverage(requiredSeconds, entries, targetKind, taskSeconds = null) {
  const tasks = normalizedTaskSeconds(requiredSeconds, taskSeconds);
  const required = tasks.reduce((sum, value) => sum + value, 0);
  const available = applicableSpeedupSeconds(entries, targetKind);
  const { remainingTaskSeconds, usedItems } = allocateWithoutOverrun(entries, targetKind, tasks);
  const remaining = remainingTaskSeconds.reduce((sum, value) => sum + value, 0);
  const applied = Math.max(0, required - remaining);
  return {
    targetKind,
    requiredSeconds: required,
    availableSeconds: available,
    appliedSeconds: applied,
    remainingSeconds: remaining,
    surplusSeconds: Math.max(0, available - applied),
    remainingTaskSeconds,
    usedItems,
  };
}

export function addPaidItemsToInventory(entries, paidItems) {
  return normalizeSpeedupInventory([
    ...(Array.isArray(entries) ? entries : []),
    ...(Array.isArray(paidItems) ? paidItems : [])
      .filter((item) => SPEEDUP_KINDS.includes(item?.kind))
      .map((item) => ({
        kind: item.kind,
        durationSeconds: integer(item.durationSeconds ?? item.duration_seconds),
        quantity: integer(item.quantity),
      })),
  ]);
}

export function paidOfferSpeedupSeconds(offer, targetKind) {
  const eligible = new Set(["general", targetKind]);
  return (offer?.items || []).reduce((total, item) => total + (
    eligible.has(item.kind)
      ? integer(item.durationSeconds ?? item.duration_seconds) * integer(item.quantity)
      : 0
  ), 0);
}

export function paidOfferGems(offer) {
  return integer(offer?.includedGems ?? offer?.included_gems)
    + integer(offer?.bonusGems ?? offer?.bonus_gems)
    + (offer?.items || []).reduce(
      (total, item) => total + (item?.kind === "gems" ? integer(item.quantity) : 0),
      0,
    );
}

function offerCoversShortfall(shortfall, _secondsPerPurchase, gemsPerPurchase, purchases, context) {
  const { offer, targetKind, taskSeconds, useGems } = context;
  const inventory = (offer?.items || [])
    .filter((item) => SPEEDUP_KINDS.includes(item?.kind))
    .map((item) => ({
      kind: item.kind,
      durationSeconds: integer(item.durationSeconds ?? item.duration_seconds),
      quantity: integer(item.quantity) * purchases,
    }));
  const coverage = speedupCoverage(shortfall, inventory, targetKind, taskSeconds);
  if (!coverage.remainingSeconds) return true;
  if (!useGems || !gemsPerPurchase) return false;
  const requiredGems = coverage.remainingTaskSeconds.reduce(
    (sum, seconds) => sum + minimumGemsForSpeedupSeconds(seconds).gems,
    0,
  );
  return gemsPerPurchase * purchases >= requiredGems;
}

function minimumOfferPurchases(shortfall, secondsPerPurchase, gemsPerPurchase, context) {
  const eligibleDurations = (context.offer?.items || [])
    .filter((item) => ["general", context.targetKind].includes(item?.kind)
      && integer(item.durationSeconds ?? item.duration_seconds) > 0
      && integer(item.quantity) > 0)
    .map((item) => integer(item.durationSeconds ?? item.duration_seconds));
  if ((!context.useGems || gemsPerPurchase <= 0)
    && eligibleDurations.length
    && context.taskSeconds.some((seconds) => !hasUnlimitedExactDuration(seconds, eligibleDurations))) {
    return null;
  }
  let high;
  if (context.useGems && gemsPerPurchase > 0) {
    const gemsWithoutSpeedups = context.taskSeconds.reduce(
      (sum, seconds) => sum + minimumGemsForSpeedupSeconds(seconds).gems,
      0,
    );
    high = Math.max(1, Math.ceil(gemsWithoutSpeedups / gemsPerPurchase));
  } else if (eligibleDurations.length) {
    const minimumDuration = Math.min(...eligibleDurations);
    high = Math.max(1, context.taskSeconds.reduce(
      (sum, seconds) => sum + Math.ceil(seconds / minimumDuration),
      0,
    ));
  } else return null;
  if (!offerCoversShortfall(shortfall, secondsPerPurchase, gemsPerPurchase, high, context)) return null;
  let low = 1;
  while (low < high) {
    const middle = Math.floor((low + high) / 2);
    if (offerCoversShortfall(shortfall, secondsPerPurchase, gemsPerPurchase, middle, context)) high = middle;
    else low = middle + 1;
  }
  return low;
}

export function recommendPaidOffers(shortfallSeconds, offers, targetKind, limit = 3, options = {}) {
  const taskSeconds = normalizedTaskSeconds(shortfallSeconds, options.taskSeconds);
  const shortfall = taskSeconds.reduce((sum, value) => sum + value, 0);
  const useGems = options.useGems === true;
  if (!shortfall) return [];
  return (Array.isArray(offers) ? offers : [])
    .map((offer) => {
      const secondsPerPurchase = paidOfferSpeedupSeconds(offer, targetKind);
      const gemsPerPurchase = paidOfferGems(offer);
      if (!secondsPerPurchase && (!useGems || !gemsPerPurchase)) return null;
      const context = { offer, targetKind, taskSeconds, useGems };
      const completingPurchases = minimumOfferPurchases(shortfall, secondsPerPurchase, gemsPerPurchase, context);
      // Keep a relevant saved offer visible even when it cannot finish every
      // task without wasting an item.  One purchase still gives the user a
      // concrete contribution and remaining-time comparison.
      const purchases = completingPurchases ?? 1;
      const totalSeconds = secondsPerPurchase * purchases;
      const availableGems = gemsPerPurchase * purchases;
      const inventory = (offer?.items || [])
        .filter((item) => SPEEDUP_KINDS.includes(item?.kind))
        .map((item) => ({
          kind: item.kind,
          durationSeconds: integer(item.durationSeconds ?? item.duration_seconds),
          quantity: integer(item.quantity) * purchases,
        }));
      const coverage = speedupCoverage(shortfall, inventory, targetKind, taskSeconds);
      const gemPurchases = coverage.remainingTaskSeconds.map(minimumGemsForSpeedupSeconds);
      const requiredGems = gemPurchases.reduce((sum, item) => sum + item.gems, 0);
      const canUseGems = useGems && coverage.remainingSeconds > 0 && availableGems >= requiredGems;
      if (!coverage.appliedSeconds && !canUseGems) return null;
      const gemsUsed = canUseGems ? requiredGems : 0;
      const gemAppliedSeconds = canUseGems ? coverage.remainingSeconds : 0;
      const remainingSeconds = Math.max(0, coverage.remainingSeconds - gemAppliedSeconds);
      const diamondCostEach = integer(offer.diamondCost ?? offer.diamond_cost);
      return {
        offerId: String(offer.offerId ?? offer.offer_id ?? ""),
        title: String(offer.title || ""),
        purchases,
        secondsPerPurchase,
        totalSeconds,
        diamondCostEach,
        totalDiamondCost: diamondCostEach ? diamondCostEach * purchases : null,
        gemsPerPurchase,
        availableGems,
        appliedSpeedupSeconds: coverage.appliedSeconds,
        gemsUsed,
        gemAppliedSeconds,
        remainingSeconds,
        excessSeconds: Math.max(0, coverage.surplusSeconds
          + gemPurchases.reduce((sum, item) => sum + item.purchasedSeconds, 0)
          - gemAppliedSeconds),
      };
    })
    .filter(Boolean)
    .sort((left, right) => Number(left.remainingSeconds > 0) - Number(right.remainingSeconds > 0)
      || Number(left.totalDiamondCost === null) - Number(right.totalDiamondCost === null)
      || (left.totalDiamondCost || 0) - (right.totalDiamondCost || 0)
      || left.excessSeconds - right.excessSeconds
      || left.purchases - right.purchases
      || left.title.localeCompare(right.title))
    .slice(0, Math.max(0, Math.trunc(limit)));
}
